optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java

```diff
@@ -737,15 +740,38 @@ public boolean hasEntityDescriptor(Class<?> entitySubclass) {
     }
 
     public EntityDescriptor<Solution_> findEntityDescriptor(Class<?> entitySubclass) {
-        return lowestEntityDescriptorMemoization.computeIfAbsent(entitySubclass, key -> {
-            // Reverse order to find the nearest ancestor
-            for (Class<?> entityClass : reversedEntityClassList) {
-                if (entityClass.isAssignableFrom(entitySubclass)) {
-                    return entityDescriptorMap.get(entityClass);
-                }
-            }
+        /*
+         * A slightly optimized variant of map.computeIfAbsent(...).
+         * computeIfAbsent(...) would require the creation of a capturing lambda every time this method is called,
+         * which is created, executed once, and immediately thrown away.
+         * This is a micro-optimization, but it is valuable on the hot path.
+         */
+        EntityDescriptor<Solution_> cachedEntityDescriptor = lowestEntityDescriptorMemoization.get(entitySubclass);
+        if (cachedEntityDescriptor == DUMMY_ENTITY_DESCRIPTOR) { // Cache hit, no descriptor found.
             return null;
-        });
+        } else if (cachedEntityDescriptor != null) { // Cache hit, descriptor found.
+            return cachedEntityDescriptor;
+        }
+        // Cache miss, look for the descriptor.
+        EntityDescriptor<Solution_> newEntityDescriptor = innerFindEntityDescriptor(entitySubclass);
+        if (newEntityDescriptor == null) {
+            // Dummy entity descriptor value, as ConcurrentMap does not allow null values.
+            lowestEntityDescriptorMemoization.put(entitySubclass, (EntityDescriptor<Solution_>) DUMMY_ENTITY_DESCRIPTOR);
+            return null;
+        } else {
+            lowestEntityDescriptorMemoization.put(entitySubclass, newEntityDescriptor);
+            return newEntityDescriptor;
+        }
```

Comment:
This code now works as a cache even when the descriptor is not found. (= null.)
Therefore when this method is called in CS during forEach(ProblemFact.class), it is significantly faster on average.